### PR TITLE
Add KyuubiMySQLFrontendService stub

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

Add KyuubiMySQLFrontendService stub, without Netty pipeline, part of #1334

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala [hunk 1]
```diff
@@ -0,0 +1,126 @@
...
+      .group(bossGroup, workerGroup)
+      .channel(classOf[NioServerSocketChannel])
+      .option(ChannelOption.SO_BACKLOG, Int.box(128))
+      .option(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
+      .childOption(ChannelOption.ALLOCATOR, PooledByteBufAllocator.DEFAULT)
+      .childOption(ChannelOption.TCP_NODELAY, Boolean.box(true))
+      .childHandler(new ChannelInitializer[SocketChannel] {
+        override def initChannel(channel: SocketChannel): Unit = channel.pipeline
+          .addLast(new LoggingHandler("org.apache.kyuubi.server.mysql.codec", LogLevel.TRACE))
+        // TODO implement authentication, codec, command handler
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:80]: Will implement in following PRs.

**Comment by pan3793** [comment]
cc @yaooqinn @yanghua @ulysses-you @timothy65535 @turboFei

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 2]
```diff
@@ -499,6 +501,67 @@ object KyuubiConf {
...
+        "Invalid thread number, must in (0, 64]")
+      .createWithDefault(1)
+
+  val FRONTEND_MYSQL_NETTY_WORKER_THREADS: ConfigEntry[Int] =
+    buildConf("frontend.mysql.netty.worker.threads")
+      .doc("Number of thread in the netty worker event loop of MySQL frontend service")
+      .version("1.4.0")
+      .intConf
+      .checkValue(n => n > 0 && n <= 64,
+        "Invalid thread number, must in (0, 64]")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:532]: why (0, 64]？

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 3]
```diff
@@ -499,6 +501,67 @@ object KyuubiConf {
...
+    .checkValue(p => p == 0 || (p > 1024 && p < 65535), "Invalid Port number")
+    .createWithDefault(3309)
+
+  val FRONTEND_MYSQL_NETTY_BOSS_THREADS: ConfigEntry[Int] =
+    buildConf("frontend.mysql.netty.boss.threads")
+      .doc("Thread number in the netty boss event loop of MySQL frontend service")
+      .version("1.4.0")
+      .intConf
+      .checkValue(n => n > 0 && n <= 64,
+        "Invalid thread number, must in (0, 64]")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:523]: why (0, 64]？

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 2 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:532]: no specific reason, it should not be too large, any experience here?

Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:532]: leave it unchecked and optional to let netty deside?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+import org.apache.kyuubi.service.{AbstractFrontendService, Serverable, Service}
+import org.apache.kyuubi.util.{NamedThreadFactory, ThreadUtils}
+
+/**
+ * A frontend service implement MySQL protocol.
+ */
+class KyuubiMySQLFrontendService(override val serverable: Serverable)
+  extends AbstractFrontendService("MySQLFrontendService") with Logging {
+
+  private var execPool: ThreadPoolExecutor = _
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:43]: where do we use it?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 2 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: Updated.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala (hunk 4 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:43]: not used yet.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala [hunk 5]
```diff
@@ -0,0 +1,126 @@
...
+  private var bootstrap: ServerBootstrap = _
+  private var bindFuture: ChannelFuture = _
+
+  @volatile protected var isStarted = false
+
+  override def initialize(conf: KyuubiConf): Unit = synchronized {
+    val poolSize = conf.get(FRONTEND_MYSQL_EXEC_POOL_SIZE)
+    val waitQueueSize = conf.get(FRONTEND_MYSQL_EXEC_WAIT_QUEUE_SIZE)
+    val keepAliveMs = conf.get(FRONTEND_MYSQL_EXEC_KEEPALIVE_TIME)
+    execPool = ThreadUtils.newDaemonQueuedThreadPool(
```
Comment (yaooqinn) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:57]: can we follow the behavior of ThriftBinaryFrontendService to use ExecutorPoolCaptureOom

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 2 above)
Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: optional?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 6]
```diff
@@ -499,6 +501,63 @@ object KyuubiConf {
...
+      .intConf
+      .createWithDefault(100)
+
+  val FRONTEND_MYSQL_EXEC_KEEPALIVE_TIME: ConfigEntry[Long] =
+    buildConf("frontend.mysql.exec.pool.keepalive.time")
+      .doc("Time(ms) that an idle async thread of the command execution thread pool will wait" +
+        " for a new task to arrive before terminating in MySQL frontend service")
+      .version("1.4.0")
+      .timeConf
+      .createWithDefault(Duration.ofSeconds(60).toMillis)
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:551]: we can follow the thrift for all of these.

On second thought, we can make those deprecated, e.g. `FRONTEND_WORKER_KEEPALIVE_TIME`  not deprecated and use these as defaults for all FEs?

**Review by yaooqinn — APPROVED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 7]
```diff
@@ -499,6 +501,63 @@ object KyuubiConf {
...
+    .intConf
+    .checkValue(p => p == 0 || (p > 1024 && p < 65535), "Invalid Port number")
+    .createWithDefault(3309)
+
+  val FRONTEND_MYSQL_NETTY_BOSS_THREADS: ConfigEntry[Int] =
+    buildConf("frontend.mysql.netty.boss.threads")
+      .doc("Thread number in the netty boss event loop of MySQL frontend service")
+      .version("1.4.0")
+      .intConf
+      .createWithDefault(1)
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:522]: option

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 8]
```diff
@@ -499,6 +501,63 @@ object KyuubiConf {
...
+
+  val FRONTEND_MYSQL_NETTY_WORKER_THREADS: ConfigEntry[Int] =
+    buildConf("frontend.mysql.netty.worker.threads")
+      .doc("Number of thread in the netty worker event loop of MySQL frontend service")
+      .version("1.4.0")
+      .intConf
+      .createWithDefault(4)
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:529]: option

**Review by yaooqinn — CHANGES_REQUESTED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 9]
```diff
@@ -499,6 +501,63 @@ object KyuubiConf {
     .checkValue(p => p == 0 || (p > 1024 && p < 65535), "Invalid Port number")
     .createWithDefault(10099)
 
+  val FRONTEND_MYSQL_BIND_HOST: OptionalConfigEntry[String] = buildConf("frontend.mysql.bind.host")
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:504]: is it better to create a new file `KyuubiMysqlConf` to define these configs ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+      try {
+        bindFuture = bootstrap.bind(serverAddr, port)
+        bindFuture.syncUninterruptibly
+        port = bindFuture.channel.localAddress.asInstanceOf[InetSocketAddress].getPort
+        info(s"MySQL frontend service has started at $connectionUrl.")
+      } catch {
+        case rethrow: Exception =>
+          throw new KyuubiException("Cannot start MySQL frontend service Netty server", rethrow)
+      }
+      isStarted = true
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:101]: inside try clause ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala [hunk 11]
```diff
@@ -0,0 +1,126 @@
...
+/**
+ * A frontend service implement MySQL protocol.
+ */
+class KyuubiMySQLFrontendService(override val serverable: Serverable)
+  extends AbstractFrontendService("MySQLFrontendService") with Logging {
+
+  private var execPool: ThreadPoolExecutor = _
+
+  var serverAddr: InetAddress = _
+  var port: Int = _
```
Comment (ulysses-you) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:46]: why they are not private ?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 8 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:529]: It's reasonable to use the `DEFAULT_EVENT_LOOP_THREADS` for worker event loop, but it is not suitable for Boss event loop.
```
DEFAULT_EVENT_LOOP_THREADS = Math.max(1, SystemPropertyUtil.getInt(
                "io.netty.eventLoopThreads", NettyRuntime.availableProcessors() * 2));
```

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 9 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:504]: Then we also need to extract thrift and rest confs

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala (hunk 11 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:46]: Will update

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala (hunk 10 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:101]: will update

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 9 above)
Comment (ulysses-you) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:504]: maybe we can do this in a new PR

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala [hunk 12]
```diff
@@ -0,0 +1,126 @@
...
+      minThreads, maxThreads,
+      keepAliveMs,
+      oomHook)
+
+    serverAddr = conf.get(FRONTEND_MYSQL_BIND_HOST)
+      .map(InetAddress.getByName)
+      .getOrElse(Utils.findLocalInetAddress)
+    port = conf.get(FRONTEND_MYSQL_BIND_PORT)
+    val workerThreads = defaultNumThreads(conf.get(FRONTEND_MYSQL_NETTY_WORKER_THREADS))
+    val bossGroup = createEventLoop(1, "mysql-netty-boss")
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala:68]: hardcode boss thread size as 1

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 13]
```diff
@@ -499,6 +502,49 @@ object KyuubiConf {
...
+    .createWithDefault(3309)
+
+  val FRONTEND_MYSQL_NETTY_WORKER_THREADS: OptionalConfigEntry[Int] =
+    buildConf("frontend.mysql.netty.worker.threads")
+      .doc("Number of thread in the netty worker event loop of MySQL frontend service")
+      .version("1.4.0")
+      .intConf
+      .checkValue(n => n > 0 && n <= MAX_NETTY_THREADS,
+        s"Invalid thread number, must in (0, $MAX_NETTY_THREADS]")
+      .createOptional
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:525]: `MAX_NETTY_THREADS` is 8

**Comment by pan3793** [comment]
Updated based on comments.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 6 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: Updated.

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala (hunk 5 above)
Comment (pan3793) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiMySQLFrontendService.scala]: Done

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 14]
```diff
@@ -499,6 +502,49 @@ object KyuubiConf {
...
+  val FRONTEND_MYSQL_BIND_PORT: ConfigEntry[Int] = buildConf("frontend.mysql.bind.port")
+    .doc("Port of the machine on which to run the MySQL frontend service.")
+    .version("1.4.0")
+    .intConf
+    .checkValue(p => p == 0 || (p > 1024 && p < 65535), "Invalid Port number")
+    .createWithDefault(3309)
+
+  val FRONTEND_MYSQL_NETTY_WORKER_THREADS: OptionalConfigEntry[Int] =
+    buildConf("frontend.mysql.netty.worker.threads")
+      .doc("Number of thread in the netty worker event loop of MySQL frontend service")
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:520]: shall we mention the default thread num is `sys.runtime.availableProcessors` in doc ?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/util/NettyUtils.scala [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  /**
+   * Returns the default number of threads for the Netty thread pools. If numUsableCores is absent,
+   * we will use Runtime get an approximate number of available cores.
+   */
+  def defaultNumThreads(numUsableCores: Option[Int]): Int = numUsableCores match {
+      case Some(num) => min(num, MAX_NETTY_THREADS)
+      case None => min(sys.runtime.availableProcessors, MAX_NETTY_THREADS)
+    }
```
Comment (ulysses-you) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/util/NettyUtils.scala:70]: nit: two more indention

**Review by ulysses-you — APPROVED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 14 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:520]: Actually, it should be `min(sys.runtime.availableProcessors, MAX_NETTY_THREADS)`, I will update it.

**Review by yaooqinn — APPROVED** [review]

File: kyuubi-common/src/main/scala/org/apache/kyuubi/util/NettyUtils.scala [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+  /**
+   * Returns the default number of threads for the Netty thread pools. If numUsableCores is absent,
+   * we will use Runtime get an approximate number of available cores.
+   */
+  def defaultNumThreads(numUsableCores: Option[Int]): Int = numUsableCores match {
+    case Some(num) => min(num, MAX_NETTY_THREADS)
+    case None => min(sys.runtime.availableProcessors, MAX_NETTY_THREADS)
+  }
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/util/NettyUtils.scala:70]: deadcode? you have already checked at KyuubiConf side?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala (hunk 13 above)
Comment (yaooqinn) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:522]: we can merge all the latter logic will one simple change `createWithDefault(MAX_NETTY_THREADS)`?

File: kyuubi-common/src/main/scala/org/apache/kyuubi/util/NettyUtils.scala (hunk 16 above)
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/util/NettyUtils.scala:70]: Just make it more generic.

**Comment by pan3793** [comment]
Thanks, merging to master
